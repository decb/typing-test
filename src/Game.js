import React from "react";
import { Button, Card, Col, Row } from "antd";

import { addWord, genWords } from "./wordsGenerator";
import Timer from "./Timer";
import UserInput from "./UserInput";
import Words from "./Words";
import { gameStates } from "./states";
import Result from "./Result";

const wordsSize = 16;

function Game(props) {
  const [gameState, setGameState] = React.useState(gameStates.BEFORE);
  const [lower, setLower] = React.useState(0);
  const [targetWords, setTargetWords] = React.useState(() =>
    genWords(props.data, wordsSize)
  );
  const [progress, setProgress] = React.useState("");
  const [inputWords, setInputWords] = React.useState([]);
  const upper = lower + wordsSize;

  const setCount = n => {
    setLower(n);
    if (upper === targetWords.length) {
      setTargetWords(addWord(props.data, targetWords));
    }
  };

  const startGame = () => setGameState(gameStates.BEFORE);
  const endGame = () => setGameState(gameStates.OVER);

  const resetGame = () => {
    setGameState(gameStates.BEFORE);
    setLower(0);
    setTargetWords(genWords(props.data, wordsSize));
    setProgress("");
    setInputWords([]);
  };

  const addInputWord = word => {
    setInputWords(inputWords.concat([word]));
    setCount(inputWords.length + 1);
  };

  const rowMargin = { marginTop: "0.8rem" };
  return (
    <Card style={{ marginTop: "0.7rem" }}>
      <Row>
        <Card
          size="small"
          title="Words to type"
          style={{ width: "100%" }}
          extra={<Timer gameState={gameState} endGame={endGame} />}
        >
          <Words
            progress={progress}
            words={targetWords}
            range={[lower, upper]}
          />
        </Card>
      </Row>
      <Row style={rowMargin}>
        <Col xs={16} sm={16} md={20} lg={20} xl={20}>
          <UserInput
            gameState={gameState}
            startGame={startGame}
            setProgress={setProgress}
            addInputWord={addInputWord}
          />
        </Col>
        <Col xs={8} sm={8} md={4} lg={4} xl={4}>
          <Button
            type="primary"
            onClick={resetGame}
            style={{
              width: "100%",
              borderTopLeftRadius: 0,
              borderBottomLeftRadius: 0
            }}
          >
            Reset
          </Button>
        </Col>
      </Row>
      {gameState === gameStates.OVER && (
        <Row style={rowMargin}>
          <Result
            inputWords={inputWords}
            targetWords={targetWords.slice(0, lower)}
          />
        </Row>
      )}
    </Card>
  );
}

export default Game;
