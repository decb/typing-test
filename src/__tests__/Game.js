import React from "react";
import renderer from "react-test-renderer";

import Game from "../Game";

const data = {
  a: [1, { a: 1 }]
};

describe("<Game />", () => {
  test("Renders correctly", () => {
    const tree = renderer.create(<Game data={data} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
