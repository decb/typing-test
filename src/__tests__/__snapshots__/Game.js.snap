// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> Renders correctly 1`] = `
<div
  className="ant-card ant-card-bordered"
  style={
    Object {
      "marginTop": "0.7rem",
    }
  }
>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <div
      className="ant-row"
      style={Object {}}
    >
      <div
        className="ant-card ant-card-bordered ant-card-small"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <div
          className="ant-card-head"
          style={Object {}}
        >
          <div
            className="ant-card-head-wrapper"
          >
            <div
              className="ant-card-head-title"
            >
              Words to type
            </div>
            <div
              className="ant-card-extra"
            >
              <span>
                60
                 seconds
              </span>
            </div>
          </div>
        </div>
        <div
          className="ant-card-body"
          style={Object {}}
        >
          <div
            style={
              Object {
                "overflow": "hidden",
                "whiteSpace": "nowrap",
              }
            }
          >
            <code>
              <span
                style={
                  Object {
                    "backgroundColor": "#ffe58f",
                    "fontWeight": "bold",
                    "padding": "3px",
                  }
                }
              >
                a
              </span>
               
              a a a a a a a a a a a a a a a
            </code>
          </div>
        </div>
      </div>
    </div>
    <div
      className="ant-row"
      style={
        Object {
          "marginTop": "0.8rem",
        }
      }
    >
      <div
        className="ant-col ant-col-xs-16 ant-col-sm-16 ant-col-md-20 ant-col-lg-20 ant-col-xl-20"
        style={Object {}}
      >
        <input
          autoCapitalize="off"
          autoComplete="off"
          autoCorrect="off"
          className="ant-input"
          disabled={false}
          onChange={[Function]}
          onKeyDown={[Function]}
          spellCheck="false"
          style={
            Object {
              "borderBottomRightRadius": 0,
              "borderTopRightRadius": 0,
            }
          }
          type="text"
          value=""
        />
      </div>
      <div
        className="ant-col ant-col-xs-8 ant-col-sm-8 ant-col-md-4 ant-col-lg-4 ant-col-xl-4"
        style={Object {}}
      >
        <button
          className="ant-btn ant-btn-primary"
          onClick={[Function]}
          style={
            Object {
              "borderBottomLeftRadius": 0,
              "borderTopLeftRadius": 0,
              "width": "100%",
            }
          }
          type="button"
        >
          <span>
            Reset
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
